### PR TITLE
Output optimal fit function after completion of fitting

### LINKED ISSUE #20326: Output optimal fit function after completion of fitting
Output the `InOut` property `Function` as one of the return values of evaluating algorithm `Fit`. This property is a python wrapper around the IFunction class (or one of its subclasses).

Useful to chain consecutive minimizations, for instance, when using different minimizers.

### PR DESCRIPTION
**Pull request by jmborr**
**To test:**

Download and unzip [test_files.zip](https://github.com/mantidproject/mantid/files/1246491/test_files.zip) under `/tmp` and run the script:

```
from __future__ import print_function
import numpy as np

resolution_file = '/tmp/BASIS_59304_divided.dat'
data_file = '/tmp/BASIS_59293_divided.dat'
resolution = LoadDaveGrp(Filename=resolution_file, OutputWorkspace='resolution', IsMicroEV=True)
data = LoadDaveGrp(Filename=data_file, OutputWorkspace='data', IsMicroEV=True)

function = '''(composite=Convolution,FixResolution=false,NumDeriv=true;
name=TabulatedFunction,Workspace=resolution,WorkspaceIndex=0,Scaling=0.5,Shift=0,XScaling=1,ties=(XScaling=1);
(name=DeltaFunction,Height=0.5,Centre=0,constraints=(0<Height<1),ties=(Centre=0);
name=Lorentzian,Amplitude=0.5,PeakCentre=0,FWHM=0.02,constraints=(0.001<FWHM),ties=(PeakCentre=0);ties=(f1.Amplitude=1-f0.Height)));
name=LinearBackground,A0=0,A1=0
'''

results = Fit(Function=function, InputWorkspace=data, WorkspaceIndex=0, CreateOutput=True)
print(results.Function)
```
The printed function should be
```
(composite=Convolution,FixResolution=false,NumDeriv=true;name=TabulatedFunction,Workspace=resolution,WorkspaceIndex=0,X=(),Y=(),Scaling=1.09281,Shift=1.47731e-10,XScaling=1,ties=(XScaling=1);(name=DeltaFunction,Height=0.87475,Centre=0,constraints=(0<Height<1),ties=(Centre=0);name=Lorentzian,Amplitude=0.12525,PeakCentre=0,FWHM=0.00874458,constraints=(0.001<FWHM),ties=(PeakCentre=0);ties=(f1.Amplitude=1-f0.Height)));name=LinearBackground,A0=3.41084e-05,A1=6.46268e-05
```

Fixes #20326
[Release Notes](https://github.com/mantidproject/mantid/blob/1d049fc61cc0b58e75dfa10546811733107a01ca/docs/source/release/v3.11.0/framework.rst)

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [x] Is the code of an acceptable quality?
- [x] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)?
- [x] Are the unit tests small and test the class in isolation?
- [x] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [x] Do changes function as described? Add comments below that describe the tests performed?
- [x] Do the changes handle unexpected situations, e.g. bad input?
- [x] Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: docs/source/algorithms/Fit-v1.rst [hunk 1]
```diff
@@ -401,16 +402,17 @@ Usage
...
    # Do the fitting
-   fitStatus, chiSq, covarianceTable, paramTable, fitWorkspace = Fit(InputWorkspace='ws', \
-      WorkspaceIndex=0, StartX = startX, EndX=endX, Output='fit', Function=myFunc)
+   fit_output = Fit(InputWorkspace='ws', WorkspaceIndex=0, StartX = startX, EndX=endX, Output='fit', Function=myFunc)
+   paramTable = fit_output.OutputParameters  # table containing the optimal fit parameters
+   fitWorkspace = fit_output.OutputWorkspace
 
-   print "The fit was: " + fitStatus
-   print("chi-squared of fit is: %.2f" % chiSq)
+   print( "The fit was: " + fit_output.OutputStatus)
```
Comment (soininen) [inline docs/source/algorithms/Fit-v1.rst]: A leftover space.

File: docs/source/algorithms/LoadSassena-v1.rst [hunk 2]
```diff
@@ -66,15 +67,16 @@ Usage
...
+    fitWorkspace = fit_output.OutputWorkspace
 
-    print "The fit was: " + fitStatus
+    print("The fit was: " + fit_output.OutputStatus)
     print("Fitted Height value is: %.2f" % paramTable.column(1)[0])
     print("Fitted centre value is: %.2f" % abs(paramTable.column(1)[1]))
     print("Fitted sigma value is: %.1f" % paramTable.column(1)[2])
     # fitWorkspace contains the data, the calculated and the difference patterns
-    print "Number of spectra in fitWorkspace is: " +  str(fitWorkspace.getNumberHistograms())
+    print( "Number of spectra in fitWorkspace is: " +  str(fitWorkspace.getNumberHistograms()))
```
Comment (soininen) [inline docs/source/algorithms/LoadSassena-v1.rst]: Leftover space.

File: docs/source/concepts/Fitting.rst [hunk 3]
```diff
@@ -179,24 +180,25 @@ implementation used for solving crystal structures from powder diffraction data.
...
+                          Minimizer='Conjugate gradient (Fletcher-Reeves imp.)')
+        costFuncVal = fit_results.OutputChi2overDoF
 
         # Here simply keep record of the best fit found, but this could easily be extended to
         # keep a record of all the minima found
         if costFuncVal < costFuncBest:
             costFuncBest = costFuncVal
             # here keep clone of best fit
-    	    CloneWorkspace(InputWorkspace='fit_Workspace',OutputWorkspace='fitBest')
+            CloneWorkspace(InputWorkspace='fit_Workspace',OutputWorkspace='fitBest')
```
Comment (soininen) [inline docs/source/concepts/Fitting.rst]: Missing space before `OutputWorkspace`.

**Review by soininen — CHANGES_REQUESTED** [review]
There were a few minor formatting issues in the usage example. Otherwise this looks like a nice addition to Fit. A small concern I have is the new API which breaks scripts that unzip the returned tuple:
```python
a, b, c, d = Fit(...)
```
However, I think we have to encourage the users to use the named tuples instead. The improvements in the usage examples are in the right direction in this regard.

**Comment by jmborr** [comment]
@soininen yes, `a, b, c, d = Fit(...)` will unfortunately break but it can't be helped. I will point out that `Fit` returns something like 7 different quantities and I think we can convince users of the futility of defining seven variables in the unpacking statement when they're planning to use, say, just one. The named tuple is definitely the way to go.

**Review by soininen — APPROVED** [review]

**Comment by mantid-roman** [comment]
I think this new property is unnecessary as it duplicates the in-out Function property and we want to get away from function strings. I would prefer if the python Fit function returned a function wrapper from Karl's PR https://github.com/mantidproject/mantid/pull/19971 instead.

**Comment by martyngigg** [comment]
@mantid-roman How would the function be retrieved after the Fit call or should it be created differently. Can you provide an example?

**Comment by mantid-roman** [comment]
I would do it similarly to how I did it for crystal field: `function = alg.getProperty('Function').value`.

**Comment by martyngigg** [comment]
@mantid-roman This is a bit harder to do in Python though isn't it?

**Comment by mantid-roman** [comment]
My example was from the crystal field python interface. `alg` is an instance of Fit algorithm. simpleapi has a wrapper for fitting algorithms. This can be done inside it.

**Comment by jmborr** [comment]
@mantid-roman I would also prefer to return a function object instead of a function string, but pull #19971 is not closed and I need this now. Also, [the returned properties by algorithms](https://github.com/mantidproject/mantid/blob/master/Framework/PythonInterface/mantid/simpleapi.py#L773) are only those marked as `Output`, so properties marked as  `InOut` are ignored.

**Comment by jmborr** [comment]
@mantid-roman If I understand, you're suggesting to modify wrapper [fitting_algorithm(f)](https://github.com/mantidproject/mantid/blob/master/Framework/PythonInterface/mantid/simpleapi.py#L299) to include returning a function object in the named tuple. Is this correct?

**Comment by mantid-roman** [comment]
@jmborr Yes

**Comment by jmborr** [comment]
OK, I'll get down to it :runner:

**Comment by jmborr** [comment]
@mantid-roman @martyngigg @soininen I'm planing of adding optional argument `inout=False` to [_gather_returns](https://github.com/mantidproject/mantid/blob/master/Framework/PythonInterface/mantid/simpleapi.py#L810) in order to retrieve not just the `Direction::Output` properties but also `Direction::InOut`, like property `Function` of the Fit algorithm. Please let me know if you have any objections.

**Comment by jmborr** [comment]
@soininen Well, this finally passed all the test. Can you please redo the manual test now that I'm returning the function object?

**Comment by soininen** [comment]
The changes look OK to me, and the test script runs just fine. However, I'm not too familiar with the fitting stuff, and how this particular PR fits the bigger picture. @mantid-roman How's your opinion?

**Comment by mantid-roman** [comment]
It fits very well. I just want to test how it works with the new `FunctionWrapper`s.

**Review by mantid-roman — CHANGES_REQUESTED** [review]
Can `Fit` return a `FunctionWrapper` instead of a C++ pointer?

**Comment by jmborr** [comment]
@mantid-roman I tried to return a `FunctionWrapper` by [invoking its initializer from the string representation of the fit function](https://github.com/mantidproject/mantid/pull/20328/files#diff-21e1642a0ce7da9cc69e0c7542489c39R887). However, I get an error from the DynamicFactory when running the test:
```
  File "/home/jbq/repositories/mantidproject/mantid2/cmake-build-debug/bin/mantid/fitfunctions.py", line 16, in __init__
    self.fun = FunctionFactory.createFunction(name)
RuntimeError: DynamicFactory: (composite=Convolution,FixResolution=false,NumDeriv=true;name=TabulatedFunction,Workspace=resolution,WorkspaceIndex=0,X=(),Y=(),Scaling=1.09281,Shift=1.47731e-10,XScaling=1,ties=(XScaling=1);(name=DeltaFunction,Height=0.87475,Centre=0,constraints=(0<Height<1),ties=(Centre=0);name=Lorentzian,Amplitude=0.12525,PeakCentre=0,FWHM=0.00874458,constraints=(0.001<FWHM),ties=(PeakCentre=0);ties=(f1.Amplitude=1-f0.Height)));name=LinearBackground,A0=3.41084e-05,A1=6.46268e-05 is not registered.
 search object (composite=Convolution,FixResolution=false,NumDeriv=true;name=TabulatedFunction,Workspace=resolution,WorkspaceIndex=0,X=(),Y=(),Scaling=1.09281,Shift=1.47731e-10,XScaling=1,ties=(XScaling=1);(name=DeltaFunction,Height=0.87475,Centre=0,constraints=(0<Height<1),ties=(Centre=0);name=Lorentzian,Amplitude=0.12525,PeakCentre=0,FWHM=0.00874458,constraints=(0.001<FWHM),ties=(PeakCentre=0);ties=(f1.Amplitude=1-f0.Height)));name=LinearBackground,A0=3.41084e-05,A1=6.46268e-05
```
I'm a bit at a loss because the new function API is, well, new to me.

File: Framework/PythonInterface/mantid/simpleapi.py [hunk 4]
```diff
@@ -862,6 +884,9 @@ def ignore_property(name_to_check, regex_to_ignore):
                     raise RuntimeError("Internal error. Output workspace property '%s' on "
                                        "algorithm '%s' has not been stored correctly."
                                        "Please contact development team." % (name,  algm_obj.name()))
+        elif _is_function_property(prop):
+            value_str = prop.valueAsStr
+            retvals[name] = FunctionWrapper(value_str)
```
Comment (mantid-roman) [inline Framework/PythonInterface/mantid/simpleapi.py]: Just pass `prop.value` to `FunctionWrapper`:

```
retvals[name] = FunctionWrapper(prop.value)
```

**Comment by jmborr** [comment]
@mantid-roman Thanks Roman, now I see your suggestion makes sense in view of [`FunctionWrapper.__init__`](https://github.com/mantidproject/mantid/blob/master/Framework/PythonInterface/mantid/fitfunctions.py#L6)

**Comment by jmborr** [comment]
@mantid-roman So as of today, is the way to create a `FunctionWrapper` object from a string representation like this?
```
ifun_obj = FunctionFactory.createInitialized("name=UserFunction,Formula=a*erf(x),a=0")
f = FunctionWrapper(ifun_obj)
```
Shouldn't `FunctionWrapper.__init__` use `FunctionFactory.createInitialized` instead of `FunctionFactory.createFunction`?

**Comment by mantid-roman** [comment]
The idea behind `FunctionWrapper` is to be able to do this:
```
f = UserFunction(Formula='a*erf(x)',a=0)
```

**Comment by jmborr** [comment]
@mantid-roman Can you please give the OK if I addressed all your concerns?

**Review by mantid-roman — APPROVED** [review]

**Comment by jmborr** [comment]
thanks! :sunflower:

**Pull request merged**